python/benchmark/gen_data_distributed.py

```diff
@@ -0,0 +1,234 @@
+            make_blobs, ["n_samples", "n_features", "return_centers"]
+        )
+        # must replace the None to the correct type
+        params["centers"] = int
+        params["random_state"] = int
+
+        return params
+
+    def gen_dataframe_and_meta(
+        self, spark: SparkSession
+    ) -> Tuple[DataFrame, List[str], np.ndarray]:
+        dtype = self.dtype
+        params = self.extra_params
+
+        if "random_state" not in params:
+            # for reproducible dataset.
+            params["random_state"] = 1
+
+        print(f"Passing {params} to make_blobs")
+
+        rows = self.num_rows
+        cols = self.num_cols
+        assert self.args is not None
+        # TODO: set num_partitions if output_num_files is not provided
+        # get default parallelism from spark context, get spark context from spark session
+        # spark.sparkContext.defaultParallelism
+        num_partitions = self.args.output_num_files
+
+        # Produce partition seeds for reproducibility.
+        random.seed(params["random_state"])
+        seed_maxval = 100 * num_partitions
+        partition_seeds = random.sample(range(1, seed_maxval), num_partitions)
+
+        partition_sizes = [rows // num_partitions] * num_partitions
+        partition_sizes[-1] += rows % num_partitions
+
+        # Generate centers upfront.
+        _, _, centers = make_blobs(
+            n_samples=0, n_features=cols, **params, return_centers=True
+        )
+
+        # Update params for partition-specific calls.
+        params["centers"] = centers
+        del params["random_state"]
+
+        # UDF to distribute make_blobs() calls across partitions. Each partition
+        # produces an equal fraction of the total samples around the predefined centers.
+        def make_blobs_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+            for pdf in iter:
+                partition_index = pdf.iloc[0][0]
+                n_partition_samples = partition_sizes[partition_index]
+                data, labels = make_blobs(
+                    n_samples=n_partition_samples,
+                    n_features=cols,
+                    **params,
+                    random_state=partition_seeds[partition_index],
+                )
+                data = np.concatenate(
+                    (
+                        data.astype(dtype),
+                        labels.reshape(n_partition_samples, 1).astype(dtype),
+                    ),
+                    axis=1,
+                )
+                yield pd.DataFrame(data=data)
+
+        label_col = "label"
+        self.schema.append(f"{label_col} {self.pyspark_type}")
+
+        return (
+            (
+                spark.range(
+                    0, num_partitions, numPartitions=num_partitions
+                ).mapInPandas(make_blobs_udf, schema=",".join(self.schema))
+            ),
+            self.feature_cols,
+            centers,
+        )
+
+
+if __name__ == "__main__":
```

Comment:
Looks like a little refactoring would allow avoiding duplication of most of this code from gen_data.py.   ie. move this code to a `main(registered_data_gens)` method in gen_data.py that allows passing in the registered data gens and then invoke (and import in this file) with the different registered_data_gens (original in gen_data.py and distributed in this file).

Reply:
The repartitioning portion is also slightly different in gen_data_distributed (we only repartition for default).